File: __Resources/Week 3/WhereAmI/WhereAmI/ViewController.m
//
//  ViewController.m
//  WhereAmI
//
//  Created by Phil Wright on 6/8/16.
//  Copyright © 2016 Touchopia, LLC. All rights reserved.
//

// 1. Import MapKit + CoreLocation Frameworks

@import MapKit;
@import CoreLocation;

#import "ViewController.h"

@interface ViewController () <MKMapViewDelegate, CLLocationManagerDelegate>

// Step 2 outlet to mapView
@property (weak, nonatomic) IBOutlet MKMapView *mapView;

// 3. Create an instance of the CLLocationManger

@property (strong, nonatomic) CLLocationManager *locationManager;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // 4. Initialize the location manager
    self.locationManager = [[CLLocationManager alloc] init];
    
    // 5. IMPORTANT set the delegate to the viewcontroller
    
    self.locationManager.delegate = self;
    
    // 6. Set the accuracy of the location manager
    self.locationManager.desiredAccuracy = kCLLocationAccuracyBest;
    
    //[self.locationManager startUpdatingLocation];
    
    //[self.locationManager startMonitoringSignificantLocationChanges];
    
    [self.locationManager requestWhenInUseAuthorization];
    
}

#pragma mark - Delegate Methods

-(void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    
    NSLog(@"Authorization changed");
    
    if (status != kCLAuthorizationStatusDenied) {
        [self.locationManager requestLocation];
    }
    
}

-(void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    NSLog(@"didFail");
}

-(void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    
    NSLog(@"didUpdate");
    
    CLLocation *location = [locations firstObject];
    CLLocationCoordinate2D coord = location.coordinate;
    
    NSLog(@"latitude == %2.8f, longitude == %2.8f", coord.latitude, coord.longitude);


    // REGION CODE TO CENTER ON MAP
    MKCoordinateRegion region = MKCoordinateRegionMake(coord, MKCoordinateSpanMake(0.01, 0.01));
    
    [self.mapView setRegion:region animated:YES];
    [self.mapView setShowsUserLocation:YES];

}

-(void)mapView:(MKMapView *)mapView didUpdateUserLocation:(MKUserLocation *)userLocation {
    
}

@end

File: 37-All-About-Maps/MapAnnotationApp/MapAnnotationApp/ViewController.swift
//
//  ViewController.swift
//  MapAnnotationApp
//
//  Created by Phil Wright on 7/10/16.
//  Copyright © 2016 Touchopia, LLC. All rights reserved.
//

import UIKit
import MapKit

class ViewController: UIViewController, MKMapViewDelegate {

    @IBOutlet weak var mapView: MKMapView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        addPin()
        //addCustomPin()
    }
 
    func addPin() {
        let location = CLLocationCoordinate2D(
            latitude: 36.169949,
            longitude: -115.13983
        )
        
        let annotation = MKPointAnnotation()
    
        annotation.coordinate = location
        annotation.title = "Las Vegas"
        annotation.subtitle = "Nevada"
        
        mapView.addAnnotation(annotation)
        
        centerMapOnLocation(location)
    }
    
    func addCustomPin() {
        let annotation = CustomPointAnnotation()
        
        let location = CLLocationCoordinate2D(
            latitude: 36.1699,
            longitude: -115.139830
        )
        
        annotation.coordinate = location
        annotation.imageName = "map"
        annotation.title = "Las Vegas"
        annotation.subtitle = "Nevada"
        
        mapView.addAnnotation(annotation)
        
    }
    
    func centerMapOnLocation(location: CLLocationCoordinate2D) {
        
        let span = MKCoordinateSpanMake(0.1, 0.1)
        let region = MKCoordinateRegion(center: location, span: span)
        mapView.setRegion(region, animated: true)
    }
    
    func mapView(mapView: MKMapView, viewForAnnotation annotation:MKAnnotation) -> MKAnnotationView?
    {
        if annotation.isKindOfClass(CustomPointAnnotation) {
    
            // 1. Create identifier
            let identifier = "MapPin"
            
            let annotationView = MKAnnotationView(annotation: annotation,reuseIdentifier: identifier)
            
            annotationView.canShowCallout = true
                
            let imageView = UIImageView(frame: CGRectMake(0,0,44,44))
            imageView.image = UIImage(named: "map")
            annotationView.image = imageView.image
            
            return annotationView
        }

        return nil
    }

}

